eurocropsml/acquisition/clipping/calibration.py

```diff
@@ -0,0 +1,170 @@
+"""Utilities for applying radiometric calibration to S-1 data."""
+
+from pathlib import Path
+from typing import Any, cast
+from xml.etree import ElementTree
+
+import numpy as np
+import xarray as xr
+import xmlschema
+
+
+def _get_xml_file(filepath: Path, band: str, identifier: str = "calibration") -> Path:
+    files: list[Path] = list(filepath.iterdir())
+    return [file for file in files if f"{band.lower()}" in str(file) and identifier in str(file)][0]
+
+
+def _parse_tag_as_list(
+    xml_path: Path,
+    query: str,
+    schema_dir: Path,
+    validation: str = "skip",
+) -> list[dict[str, Any]]:
+    """Function to parse xml tags into list.
+
+    Adjusted from xarray-sentinel (https://github.com/bopen/xarray-sentinel).
+    """
+    schema = xmlschema.XMLSchema(schema_dir)
+    xml_tree = ElementTree.parse(xml_path)
+    tag: Any = schema.decode(xml_tree, query, validation=validation)
+    if tag is None:
+        tag = []
+    elif isinstance(tag, dict):
+        tag = [tag]
+    tag_list: list[dict[str, Any]] = tag
+    assert isinstance(tag_list, list), f"{type(tag_list)} is not list"
+    return tag_list
+
+
+def _open_noise_dataset(safe_dir: Path, band: str) -> xr.Dataset:
+    """Function to read noise from LUT.
+
+    This reads the NADS (Noise Annotation Data Set) file.
+
+    Adjusted from xarray-sentinel (https://github.com/bopen/xarray-sentinel).
+    """
+    xml_dir: Path = safe_dir / "annotation" / "calibration"
+    schema_dir: Path = safe_dir / "support" / "s1-level-1-noise.xsd"
+    xml_calibration_file = _get_xml_file(xml_dir, band, identifier="noise")
+    noise_vectors = _parse_tag_as_list(xml_calibration_file, ".//noiseRangeVector", schema_dir)
+
+    pixel_list = []
+    line_list = []
+    noise_range_lut_list = []
+    for vector in noise_vectors:
+        line_list.append(vector["line"])
+        pixel = np.fromstring(vector["pixel"]["$"], dtype=int, sep=" ")
+        pixel_list.append(pixel)
+        noise_range_rut = np.fromstring(vector["noiseRangeLut"]["$"], dtype=np.float32, sep=" ")
+        noise_range_lut_list.append(noise_range_rut)
+
+    pixel = np.array(pixel_list)
+    if (pixel - pixel[0]).any():
+        raise ValueError("Unable to organize noise vectors in a regular line-pixel grid")
+    data_vars = {
+        "noiseRangeLut": (("line", "pixel"), noise_range_lut_list),
+    }
+    coords = {"line": line_list, "pixel": pixel_list[0]}
+
+    return xr.Dataset(data_vars=data_vars, coords=coords)
+
+
+def _open_calibration_dataset(safe_dir: Path, band: str) -> xr.Dataset:
+    """Function to read calibration from LUT.
+
+    This reads the CADS (Calibration Annotation Data Set) file.
+
+    Adjusted from xarray-sentinel (https://github.com/bopen/xarray-sentinel).
+    """
+    xml_dir: Path = safe_dir / "annotation" / "calibration"
+    schema_dir: Path = safe_dir / "support" / "s1-level-1-calibration.xsd"
```

Comment:
How would the users get access the calibration file? Is the plan to include this in the repo or the Zenodo data record?

Reply:
These files are part of the Sentinel-1 `.SAFE` files which also include the actual Sentinel-1 data (`GeoTIFFs`), cf. [documentation](https://sentinel.esa.int/documents/247904/685163/S1-Radiometric-Calibration-V1.0.pdf). So the user needs access to them anways. Either through EO-Lab or any other platform that provides Sentinel `.SAFE` files.